### PR TITLE
Stability fixes

### PR DESCRIPTION
**Pull request by wutschel**
Close many issues reported in https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/193.

- Avoid out-of-bounds access at several places reported from crash logs
- Enhance thread safety for EPG which was suspect to cause several crashes in this area
- Avoid crashes in some other places

### CODE REVIEW

File: XBMC Remote/NowPlaying.m [hunk 1]
```diff
@@ -692,11 +692,11 @@ -(void)getActivePlayers{
                          bool enableJewel = [self enableJewelCases];
                          if( [NSJSONSerialization isValidJSONObject:methodResult]){
                              NSDictionary *nowPlayingInfo = [methodResult objectForKey:@"item"];
-                             if ([nowPlayingInfo  objectForKey:@"id"] == nil)
+                             if (((NSNull *)[nowPlayingInfo objectForKey:@"id"] == [NSNull null]))
```
Comment (kambala-decapitator) [inline XBMC Remote/NowPlaying.m]: not sure why you're casting to `NSNull` - does compiler complain when comparing `id` and `NSNull`?

File: XBMC Remote/NowPlaying.m [hunk 2]
```diff
@@ -662,12 +662,9 @@ -(void)getActivePlayers{
...
             if( [methodResult count] > 0){
                 nothingIsPlaying = NO;
-                NSNumber *response;
-                if (((NSNull *)[[methodResult objectAtIndex:0] objectForKey:@"playerid"] != [NSNull null])){
-                    response = [[methodResult objectAtIndex:0] objectForKey:@"playerid"];
-                }
-                currentPlayerID=[response intValue];
-                if (playerID!=[response intValue] || (selectedPlayerID>-1 && playerID!=selectedPlayerID)){  // DA SISTEMARE SE AGGIUNGONO ITEM DALL'ESTERNO: FUTURA SEGNALAZIONE CON SOCKET!                    
+                NSNumber *response = methodResult[0][@"playerid"];
+                currentPlayerID = [response intValue];
```
Comment (kambala-decapitator) [inline XBMC Remote/NowPlaying.m:666]: `NSNull` comparison got lost. Is it a correct change? (as I see the same below as well)

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]
- 7a560a72ebdff8aa9f0bb308f02fcae04614f7cd - I suggest to put it in your "code modernising" branch instead
- 5c5f0569176b993efc7a05bcd2dff362b374e568 - first, I'd avoid `NSLock` completely due to this warning from docs and use `dispatch_queue`: "The NSLock class uses POSIX threads to implement its locking behavior. When sending an unlock message to an NSLock object, you must be sure that message is sent from the same thread that sent the initial lock message. Unlocking a lock from a different thread can result in undefined behavior." Second, it looks that you're protecting access to 2 different entities (`epgDict` and `epgDownloadQueue`) with the same lock which doesn't look correct.

**Comment by wutschel** [comment]
> * [7a560a7](https://github.com/xbmc/Official-Kodi-Remote-iOS/commit/7a560a72ebdff8aa9f0bb308f02fcae04614f7cd) - I suggest to put it in your "code modernising" branch instead

Sure, can do this. Will remove the commit when refreshing this PR.

> * [5c5f056](https://github.com/xbmc/Official-Kodi-Remote-iOS/commit/5c5f0569176b993efc7a05bcd2dff362b374e568) - first, I'd avoid `NSLock` completely due to this warning from docs and use `dispatch_queue`: "The NSLock class uses POSIX threads to implement its locking behavior. When sending an unlock message to an NSLock object, you must be sure that message is sent from the same thread that sent the initial lock message. Unlocking a lock from a different thread can result in undefined behavior." Second, it looks that you're protecting access to 2 different entities (`epgDict` and `epgDownloadQueue`) with the same lock which doesn't look correct.

Ok, will look into using `dispatch_queue` then. I am protecting both using the same lock as I felt using two separate locks is overcomplicating things. It is easier the use a common lock to protect this "topic".

File: XBMC Remote/NowPlaying.m (hunk 2 above)
Comment (wutschel) [reply XBMC Remote/NowPlaying.m:666]: Yes, to my knowledge this does the same. Before, `response` was left unitialized (= `nil`), now it gets `nil` assigned when the former condition is not met.

**Comment by kambala-decapitator** [comment]
> I am protecting both using the same lock as I felt using two separate locks is overcomplicating things. It is easier the use a common lock to protect this "topic".

it'd mean that access to one blocks the other. Would it be correct?

File: XBMC Remote/NowPlaying.m (hunk 2 above)
Comment (kambala-decapitator) [reply XBMC Remote/NowPlaying.m:666]: No, it's not the same. `NSNull` is an object while `nil` is nothing. Sending random messages to the latter is ok (evaluates to 0 / NO / nil) while sending unsupported messages to objects results in crash. Consider the following identical piece of code:

```objc
NSNumber *foo = (id)[NSObject new];
[foo intValue]; // crash
```

Comment (wutschel) [reply XBMC Remote/NowPlaying.m:666]: Thanks, got it now. Will throw the commit away for now.

File: XBMC Remote/NowPlaying.m (hunk 1 above)
Comment (wutschel) [reply XBMC Remote/NowPlaying.m]: I guess then I was completely off track with what I wanted to achieve. In the crash logs I see a crashes reported for the line

`if ([nowPlayingInfo  objectForKey:@"id"] == nil)`

with following call stack:

```
Last Exception Backtrace:
0   CoreFoundation                	0x18ee9286c __exceptionPreprocess + 220 (NSException.m:199)
1   libobjc.A.dylib               	0x1a3eadc50 objc_exception_throw + 60 (objc-exception.mm:565)
2   CoreFoundation                	0x18ed9995c -[NSObject(NSObject) doesNotRecognizeSelector:] + 144 (NSObject.m:146)
3   CoreFoundation                	0x18ee95438 ___forwarding___ + 1444 (NSForwarding.m:3597)
4   CoreFoundation                	0x18ee97740 _CF_forwarding_prep_0 + 96
5   Kodi Remote                   	0x101034a50 __30-[NowPlaying getActivePlayers]_block_invoke_2 + 184 (NowPlaying.m:695)
```

From forum posts I got a wrong understanding of what might have happened. Now I understand problem could come from `objectForKey` called on a wrong type of `nowPlayingInfo` -- even though having the correct type -- might have the wrong content.

Would in such case checking for something like `[methodResult[@"item"] isKindOfClass:[NSDictionary class]]` or also a check for `![NSNull null]` help?

**Comment by wutschel** [comment]
> it'd mean that access to one blocks the other. Would it be correct?

In the function `saveEPGToDisk` both arrays are used. I was aiming for consistent data and keep both arrays in the same lock.

**Comment by wutschel** [comment]
Rebased and threw out the changes in `NowPlaying.m` which were in doubt. Still looking into replacing `NSLock`.

**Comment by wutschel** [comment]
Now replaced `NSLock` with `dispatch_queue`.

File: XBMC Remote/NowPlaying.m (hunk 1 above)
Comment (wutschel) [reply XBMC Remote/NowPlaying.m]: Done

File: XBMC Remote/DetailViewController.m [hunk 3]
```diff
@@ -5565,6 +5556,7 @@ - (void)viewDidLoad{
     isViewDidLoad = YES;
     sectionHeight = 16;
     dataList.tableFooterView = [UIView new];
+    epglockqueue = dispatch_queue_create("com.epg.arrayupdate", DISPATCH_QUEUE_CONCURRENT);
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/DetailViewController.m]: Thanks for poking me on this. I misread the documentation. :/

File: XBMC Remote/DetailViewController.m [hunk 4]
```diff
@@ -128,25 +128,23 @@ - (NSMutableDictionary *) indexKeyedMutableDictionaryFromArray:(NSArray *)array
 #pragma mark - live tv epg memory/disk cache management
 
 -(NSMutableArray *)loadEPGFromMemory:(NSNumber *)channelid {
-    return [epgDict objectForKey:channelid];
+    __block NSMutableArray *epgarray = nil;
+    dispatch_barrier_async(epglockqueue, ^{
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/DetailViewController.m]: Consequence of me misreading the doc. When using serial queue `dispatch_async` is the right choice.

**Comment by wutschel** [comment]
Changed EPG synchronization to `DISPATCH_QUEUE_SERIAL` and `display_sync`. Rebased to master.

**Pull request merged**